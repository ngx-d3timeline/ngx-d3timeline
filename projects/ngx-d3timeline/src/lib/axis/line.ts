import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
