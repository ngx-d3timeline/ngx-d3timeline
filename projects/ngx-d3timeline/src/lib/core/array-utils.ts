export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}
