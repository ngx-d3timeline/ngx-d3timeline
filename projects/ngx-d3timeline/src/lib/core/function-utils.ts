export function partialApply<TInput, TAppliedArgs extends any[], TOut>(
  func: (input: TInput, ...appliedArgs: TAppliedArgs) => TOut
): (...appliedArgs: TAppliedArgs) => (input: TInput) => TOut {
  return (...appliedArgs: TAppliedArgs) => (input: TInput) =>
    func(input, ...appliedArgs);
}

export function pipe<T>(
  ...args: ((...args: any[]) => T)[]
): (...t: any[]) => T {
  return x => args.reduce((acc, curr) => curr(acc), x);
}

export function clampZero(x: number) {
  return Math.max(x, 0);
}
