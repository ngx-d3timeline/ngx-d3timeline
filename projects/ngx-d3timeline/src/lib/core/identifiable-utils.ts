import { Identifiable, identifier } from './types';

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
