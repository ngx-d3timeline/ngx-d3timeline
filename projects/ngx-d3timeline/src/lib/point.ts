import { Orientation } from './orientation';

export interface Point {
  x: number;
  y: number;
}

export const origin: Point = {
  x: 0,
  y: 0
};

export function translatePoint(point: Point, offset: Point): Point {
  return { x: point.x + offset.x, y: point.y + offset.y };
}

export function translatePointInOrientation(
  point: Point,
  distance: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? { ...point, y: point.y + distance }
    : { ...point, x: point.x + distance };
}
