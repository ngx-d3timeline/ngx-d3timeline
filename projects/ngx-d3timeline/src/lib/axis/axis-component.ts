import { Component, Input } from '@angular/core';
import { Axis } from './axis';

@Component({
  selector: '[ngx-d3timeline-axis]',
  template: `
    <svg:g class="axis-group">
      <svg:g
        *ngFor="let tickMark of axis.tickMarks"
        ngx-d3timeline-axis-tick-mark
        [tickMark]="tickMark"
      ></svg:g>
      <svg:g ngx-d3timeline-line [line]="axis.line"></svg:g>
    </svg:g>
  `
})
export class AxisComponent {
  @Input() axis: Axis;
}
