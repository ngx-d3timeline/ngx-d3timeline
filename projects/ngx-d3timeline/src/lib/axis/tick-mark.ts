import { Line } from './line';
import { Point } from '../point';

export interface TickMark {
  label: string;
  labelOffset: Point;
  transform: string;
  line: Line;
}
