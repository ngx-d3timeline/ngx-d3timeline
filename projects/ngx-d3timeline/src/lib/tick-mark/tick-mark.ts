import { Line } from '../core/line';
import { Point } from '../core/point';

export interface TickMark {
  label: string;
  labelOffset: Point;
  transform: string;
  line: Line;
}
