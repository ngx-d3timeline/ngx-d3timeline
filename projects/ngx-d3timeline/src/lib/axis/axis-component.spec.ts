import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AxisComponent } from './axis-component';
import { createLine, Line } from './line';
import { origin } from '../point';
import { Component, Input } from '@angular/core';
import { TickMark } from './tick-mark';

@Component({
  selector: '[ngx-d3timeline-line]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeLineComponent {
  @Input() line: Line;
}

@Component({
  selector: '[ngx-d3timeline-axis-tick-mark]',
  template: `
    <svg:g class="axis-tick-mark"></svg:g>
  `
})
class FakeAxisTickMarkComponent {
  @Input() tickMark: TickMark;
}

describe('AxisComponent', () => {
  let fixture: ComponentFixture<AxisComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        AxisComponent,
        FakeLineComponent,
        FakeAxisTickMarkComponent
      ]
    });

    fixture = TestBed.createComponent(AxisComponent);
  });

  it('should render correctly', () => {
    const line = createLine(origin, { x: 10, y: 10 });
    fixture.componentInstance.axis = {
      tickMarks: [
        {
          label: 'tick 1',
          labelOffset: { x: 0, y: 1 },
          transform: 'translate(0, 10)',
          line
        },
        {
          label: 'tick 2',
          labelOffset: { x: 0, y: 2 },
          transform: 'translate(0, 20)',
          line
        }
      ],
      line: { x1: 0, x2: 10, y1: 1, y2: 0 }
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
