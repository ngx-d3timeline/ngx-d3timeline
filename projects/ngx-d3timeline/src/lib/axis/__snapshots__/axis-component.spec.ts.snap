// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxisComponent should render correctly 1`] = `
<div
  id="root0"
  ngx-d3timeline-axis=""
>
  <g
    class="axis-group"
  >
    
    <g
      ngx-d3timeline-axis-tick-mark=""
    >
      <g
        class="axis-tick-mark"
      />
    </g>
    <g
      ngx-d3timeline-axis-tick-mark=""
    >
      <g
        class="axis-tick-mark"
      />
    </g>
    <g
      ngx-d3timeline-line=""
    >
      <g />
    </g>
  </g>
</div>
`;
